### PR TITLE
Feature/use checkout hook

### PR DESCRIPTION
**Pull request by Rodriguespn**
<h2>Greptile Overview</h2>

### Greptile Summary

This PR adds Instant Checkout support for ChatGPT apps through a new `useCheckout` hook, following the ACP specification. The implementation includes comprehensive TypeScript types, thorough test coverage, and good documentation.

**Major changes:**
- New `useCheckout` hook for payment processing with `window.openai.requestCheckout`
- Refactored `useCallTool` to use shared `useAsyncOperation` hook for state management
- Added `useAsyncOperation` hook to extract common async operation patterns with deduplication support
- Complete TypeScript type definitions for checkout sessions, responses, and errors
- Comprehensive test suites for both new hooks (443 lines for checkout, 411 lines for async operations)
- Detailed API documentation with multiple usage examples

**Issues found:**
- Minor state inconsistency risk with `sessionId` when deduplication is enabled
- Unrelated change in `widgetsDevServer.ts` that removes unused variable destructuring

The refactoring is well-executed, consolidating duplicate state management logic into a reusable hook. The checkout implementation follows existing patterns from `useCallTool`, providing both callback-based and async/await APIs.

### Confidence Score: 4/5

- This PR is safe to merge with minor considerations for the identified state management edge case
- The implementation is well-designed with comprehensive tests, clear documentation, and follows established patterns in the codebase. The refactoring consolidates duplicate code effectively. The identified sessionId state inconsistency is a minor edge case that would only manifest under rapid sequential checkout calls, which is unlikely in practice. The unrelated widgetsDevServer change is cosmetic and unlikely to cause issues.
- The `use-checkout.ts` file could benefit from addressing the sessionId state management if handling rapid sequential checkouts is a requirement

### CODE REVIEW

**Comment by Rodriguespn** [comment]
Hey @Fred! Thanks for the feedback and patience. I've addressed all your review points you mentioned besides the server tool implementation for ACP implementation:

- Code Consolidation: Created a shared `useAsyncOperation` base hook that extracts the common async state management pattern. Both `useCheckout` and `useCallTool` now use this base, which eliminated around 80-100 lines of duplicated code. The refactor maintains full backward compatibility with all 95 tests passing.

- Checkout-Specific Features: Added automatic session ID generation with `crypto.randomUUID()` that's overridable via the `checkoutSessionIdGenerator` option. The `sessionId` is now exposed directly in the hook's return value. Also changed the response to return `order` directly instead of nesting it in a `data` attribute. Added 7 new tests covering session ID generation and order access patterns.

- Documentation: Removed all JSDoc and in-file comments from hook files. Improved the `use-checkout.md` documentation with comprehensive examples including session ID generation and a full API reference.

- Maintenance: Upgraded biome from `2.3.10` to `2.3.11` to align with the version of the dev dependency on `package.json`. Also removed unused variables in packages/core/src/server/widgetsDevServer.ts that were causing format issues when running `pnpm run test:format`. Happy to revert these changes or open a new PR for them.

Removed the example app since is out of scope, removed in-file comments and moved them to proper docs, consolidated the shared codebase with the new base hook, and implemented the checkout-specific features you mentioned. 

Regarding the ACP implementation for OpenAI's server tool, happy to tackle that in a separate PR later as you suggested.

File: docs/docs/api-reference/use-checkout.md [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+    </div>
+  );
+}
+```
+
+## Parameters
+
+```tsx
+const {
+  data,
```
Comment (fredericbarthelet) [inline docs/docs/api-reference/use-checkout.md:39]: Do you still expose data in addition to order? I feel like this might be redundant

File: docs/docs/api-reference/use-checkout.md [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+function CheckoutStatus({ session }: { session: CheckoutSessionRequest }) {
+  const { requestCheckout, isPending, isSuccess, order, sessionId } = useCheckout();
+
+  return (
+    <div>
+      <button disabled={isPending} onClick={() => requestCheckout(session)}>
+        {isPending ? "Processing..." : "Checkout"}
+      </button>
+      {sessionId && <p>Session ID: {sessionId}</p>}
+      {isSuccess && order && (
```
Comment (fredericbarthelet) [inline docs/docs/api-reference/use-checkout.md:337]:
```suggestion
      {isSuccess && (
```
I feel like isSuccess should be sufficient for order to be defined

File: packages/core/src/server/widgetsDevServer.ts [hunk 3]
```diff
@@ -26,7 +26,7 @@ export const widgetsDevServer = async (): Promise<RequestHandler> => {
     webAppRoot,
   );
 
-  const { build, preview, ...devConfig } = configResult?.config || {};
+  const { ...devConfig } = configResult?.config || {};
```
Comment (fredericbarthelet) [inline packages/core/src/server/widgetsDevServer.ts:29]: I think you might have a conflicting version of biome on your IDE (or at least different that the one being used on this project). We experienced a few issues recently on that front: the extreme step of re downloading the project from Github did the trick for most of the team.

Anyhow, this should not be part of this PR

File: packages/core/src/web/hooks/use-async-operation.test.ts [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+
+  describe("successful execution", () => {
+    it("should transition from idle -> pending -> success", async () => {
+      const { result } = renderHook(() => useAsyncOperation<string>());
+      const mockData = "success data";
+
+      let resolveOperation: (value: string) => void;
+      const operationPromise = new Promise<string>((resolve) => {
+        resolveOperation = resolve;
+      });
```
Comment (fredericbarthelet) [inline packages/core/src/web/hooks/use-async-operation.test.ts:32]: You have a very convenient Promise syntax to create such "to be resolved" promise: [`Promise.withResolvers`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers)

File: packages/core/src/web/hooks/use-async-operation.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+export type AsyncOperationState<TData, TError = unknown> =
+  | AsyncOperationIdleState<TData, TError>
+  | AsyncOperationPendingState<TData, TError>
+  | AsyncOperationSuccessState<TData, TError>
+  | AsyncOperationErrorState<TData, TError>;
+
+export type UseAsyncOperationConfig = {
+  enableDeduplication?: boolean;
+};
```
Comment (fredericbarthelet) [inline packages/core/src/web/hooks/use-async-operation.ts:51]: All the implementations of this hook seem to enable deduplication. I feel like this config parameter is not useful

File: packages/core/src/web/hooks/use-async-operation.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  status: "error";
+  isIdle: false;
+  isPending: false;
+  isSuccess: false;
+  isError: true;
+  data: undefined;
+  error: TError;
+};
+
+export type AsyncOperationState<TData, TError = unknown> =
```
Comment (fredericbarthelet) [inline packages/core/src/web/hooks/use-async-operation.ts:43]: There is not a single location where AsyncOperationState is defined without its generic types, so you can drop the default generic value

```suggestion
export type AsyncOperationState<TData, TError> =
```

**Comment by fredericbarthelet** [comment]
@greptile could you review again?

File: packages/core/src/web/hooks/use-call-tool.ts [hunk 7]
```diff
@@ -103,38 +60,22 @@ export const useCallTool = <
...
-      }
-
-      return data;
-    } catch (error) {
-      if (callId === callIdRef.current) {
-        setCallToolState({ status: "error", data: undefined, error });
-      }
-      throw error;
-    }
+    });
```
Comment (fredericbarthelet) [inline packages/core/src/web/hooks/use-call-tool.ts:78]: why not use `execute` returned by `useAsyncOperation` directly and instead redefining a monad here?

Comment (fredericbarthelet) [reply packages/core/src/web/hooks/use-call-tool.ts:78]: Maybe it will be easier to understand if `useAsyncOperation` takes in the async callback as argument instead?

```ts
const { state, execute } = useAsyncOperation<
    CombinedCallToolResponse,
    unknown
  >((toolArgs) => adaptor.callTool(name, toolArgs));
```

File: packages/core/src/web/hooks/use-checkout.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    });
+  };
+
+  const requestCheckoutAsync: RequestCheckoutAsyncFn = (
+    session: CheckoutSessionRequest,
+  ) => {
+    return execute(session);
+  };
+
+  const requestCheckout: RequestCheckoutFn = ((
```
Comment (fredericbarthelet) [inline packages/core/src/web/hooks/use-checkout.ts:105]: can't this be part of the `useAsyncOperation` as well? I feel like it's duplicated as well

**Review by fredericbarthelet — CHANGES_REQUESTED** [review]
Hey @Rodriguespn thanks for your contribution. It's close to be ready!
I just shared a few comments to improve tests suite and `useAsyncOperation` hook. Let me know what you think of it
